standard_configure.sh

```diff
@@ -1,9 +1,141 @@
+#
+# MACOSX_DEPLOYMENT_TARGET
+#   The minimum macOS version on which the compiled libraries and programs are
+#   intended to run. It has not been tested whether building for an earlier
+#   deployment target results in all aspects of MacPorts functioning correctly
+#   on earlier macOS versions, so changing the value of this variable is not
+#   recommended except to perform such testing. In addition, some values that
+#   vary based on macOS version are recorded in text files like macports.conf,
+#   macports_autoconf.tcl and port_autoconf.tcl, so they must be generated on
+#   the same major version of macOS as the one on which they will be used.
+#
+# OPTFLAGS
+#   Compiler optimization flags.
+#
+# SDKPATH
+#   The path to the macOS SDK with which to build.
+#
+# UNIVERSAL
+#   Whether the default ARCHS are universal (yes/no).
 
-# If you want to use a different prefix, add this to the above:
-# --prefix=/some/path --with-applications-dir=/some/path/Applications
+set -eu
+
+SCRIPT_DIR=$(cd "$(dirname "$0")" && pwd)
+MACPORTS_VERSION=$(cat "$SCRIPT_DIR/config/macports_version")
+CONFIGURE=$SCRIPT_DIR/configure
+DEFAULT_CFLAGS=-pipe
+DEFAULT_LDFLAGS=
+PATH=/usr/bin:/bin:/usr/sbin:/sbin
+
+SYSTEM_NAME=$(uname -s)
+if [ "$SYSTEM_NAME" = "Darwin" ]; then
+    DEFAULT_LDFLAGS="$DEFAULT_LDFLAGS -Wl,-headerpad_max_install_names"
+    MACOS_VERSION_MAJOR=$(sw_vers -productVersion | cut -d. -f1-2)
+    : "${MACOSX_DEPLOYMENT_TARGET=$MACOS_VERSION_MAJOR}"
+    case $MACOS_VERSION_MAJOR in
+        10.[0123])
+            echo "MacPorts $MACPORTS_VERSION requires Mac OS X 10.4 or later." 1>&2
+            exit 1
+            ;;
+        10.4)
+            : "${CC=/usr/bin/gcc-4.0}"
+            : "${SDKPATH=/Developer/SDKs/MacOSX10.4u.sdk}"
+            ;;
+        10.5)
+            : "${CC=/usr/bin/gcc-4.2}"
+            ;;
+        *)
+            : "${CC=/usr/bin/clang}"
+            ;;
+    esac
+    case $MACOS_VERSION_MAJOR in
+        10.[456])
+            : "${UNIVERSAL=yes}"
+            ;;
+        *)
+            : "${UNIVERSAL=no}"
+            ;;
+    esac
+    case $MACOS_VERSION_MAJOR in
+        10.[45])
+            if [ "$UNIVERSAL" = "yes" ]; then
+                : "${ARCHS=ppc i386}"
+            else
+                if [ "$(uname -m)" = "Power Macintosh" ]; then
+                    : "${ARCHS=ppc}"
+                else
+                    : "${ARCHS=i386}"
+                fi
+            fi
+            ;;
+        *)
+            if [ "$UNIVERSAL" = "yes" ]; then
+                : "${ARCHS=x86_64 i386}"
+            else
+                if [ "$(sysctl -n hw.cpu64bit_capable)" = "1" ]; then
+                    : "${ARCHS=x86_64}"
+                else
+                    : "${ARCHS=i386}"
+                fi
+            fi
+            ;;
+    esac
+fi
+
+if [ -z "${ARCHFLAGS-}" ]; then
+    for A in ${ARCHS-}; do
+        ARCHFLAGS="${ARCHFLAGS-} -arch $A"
+    done
+    ARCHFLAGS="${ARCHFLAGS# }"
+fi
+
+: "${CC=cc}"
```

Comment:
Shouldn't we let the configure script figure this out? It is currently checking for clang, cc, gcc, in this order and would therefore prefer `clang` over `cc`.

Reply:
We could let the configure script figure it out, at the expense of not allowing a user to override it by environment variable when calling standard_configure.sh. Or we could pass along a user-provided `CC` if set, and otherwise not pass along any value.

My script also defaults to `clang` on 10.6 and up, you'll notice. So that wouldn't be a difference.

My script uses `gcc-4.2` on 10.5, like the MacPorts port does (because that's the MacPorts base default compiler on 10.5). If we let the configure script use `cc`, that would be `gcc-4.0`, because that was Apple's default compiler on 10.5. The reason why MacPorts base defaults to `clang` on 10.6 was presumably to use the newest and best compiler available, so I wanted to use the newest and best compiler available on 10.5 too. Maybe that change belongs in base rather than in this script.

On the other hand, I'm using `gcc-4.0` on 10.4, unlike the MacPorts port which uses `gcc-apple-4.2` because that's the MacPorts base default compiler on 10.4. I didn't want to require that the apple-gcc42 port be installed.

Presumably, as soon as the user runs `sudo port selfupdate`, base gets rebuilt using whatever compiler the configure script finds, so any advantage of having the binary built with a "better" compiler is lost at that point.

Should selfupdate run standard_configure.sh?

Reply:
It looks like selfupdate sets `CC=/usr/bin/cc OBJC=/usr/bin/cc`. Should it not do that, and let the configure script decide?

Reply:
> It looks like selfupdate sets `CC=/usr/bin/cc OBJC=/usr/bin/cc`. Should it not do that, and let the configure script decide?

We're doing it because [8 years ago I said we should](https://trac.macports.org/ticket/23095). But that presumes we allow the user's `PATH` to be used. Do we? I hope not. If we changed selfupdate to use standard_configure.sh, that would solve that.

Reply:
`PATH` will be set up by mportinit before running the selfupdate code. I assume `/usr/bin/cc` never caused problems because it already pointed to clang on macOS versions where we wanted to prefer clang (although it might have pointed to llvm-gcc-4.2 at some point?).

We should put the logic to prefer `gcc-4.2` on 10.5 and `gcc-4.0` on 10.4 into the configure script if we always want that. But that also seems to be the case for the other settings here. Therefore, instead of requiring `standard_configure.sh` to be run, we could move these defaults directly into `configure`. However, I would see that as a long-term goal and it is not a showstopper for this pull request.

Reply:
On 10.6, `/usr/bin/cc` is gcc-4.2, not clang. Using gcc-4.2 doesn't cause any problems that I'm aware of, but it contradicts the intention of ce44cdcc04b59d0b4f08038b7096d616280110e1 which made the configure script prefer clang (though there was no explanation for why that change was made; we could revert it).

On 10.4, `/usr/bin/cc` is gcc-4.0, so there's no need to do anything special there to use that compiler.

standard_configure.sh was created in 70d6bdb6062b88f5f13c3ee754f957b18f7d13b8 so that we had a place to specify that for a "standard" build, we want to ignore any custom `PATH` and that we want to use `--enable-readline`, even though there were reasons why we did not want the configure script to do those things by default. Perhaps similar reasoning applies to some of the settings I'm proposing in this PR. Or perhaps not. I haven't thought about it a great deal.

My goal with this PR was to move functionality that is currently implemented in the MacPorts port into MacPorts base. It did not occur to me to attempt to modify the configure script. I assumed there were good reasons why we weren't doing these things in the configure script already.

Reply:
An additional issue is that although our configure script prefers clang, it does not pass that preference on to the other configure scripts it runs, like tcl's. This means that on 10.6, if we don't set `CC` when running our configure, MacPorts base is built with clang but tcl and the other bundled packages are built with gcc. It seems better to build everything we ship with the same compiler and same flags. Setting then in standard_configure.sh accomplishes that. It may be possible to accomplish the same through modifications of the configure script but I have not looked into it.

Reply:
I removed the unused `OBJC` with 8475592e and 9f8eb1d8.

See #89 for a way to pass CC to configure in vendor sources. If it makes things easier, you can also fold the changes into this pull request and close #89, but I did not want to just push them to this branch right away as the changes do not actually depend on each other.

Reply:
#89 doesn't seem related to #86; I would keep them separate.